### ISSUE TITLE
Implement np.shape for nestedarray

### ISSUE THREAD

**Issue by astoeriko** [numba/numba#7526]
## Feature request

numba currently (0.54.1) doesn't implement `np.shape` for nestedarrays from numpy record types. Support for this would simplify some generic code a little bit.

```python
import numba
import numpy as np

dtype = np.dtype([
    ("a", np.float64, (10,))
])

data = np.zeros((), dtype=dtype)[()]

@numba.njit
def foo(data):
    #return data.a.shape  # this works
    return np.shape(data.a)

foo(data)  # Throws a TypeError
```

<details>

```
TypingError: Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function(<function shape at 0x7fb0b028e430>) found for signature:
 
 >>> shape(nestedarray(float64, (10,)))
 
There are 2 candidate implementations:
  - Of which 2 did not match due to:
  Overload in function 'np_shape': File: numba/np/arrayobj.py: Line 1981.
    With argument(s): '(nestedarray(float64, (10,)))':
   Rejected as the implementation raised a specific error:
     TypingError: Failed in nopython mode pipeline (step: nopython frontend)
   No implementation of function Function(<function asarray at 0x7fb0b026a040>) found for signature:
    
    >>> asarray(nestedarray(float64, (10,)))
    
   There are 2 candidate implementations:
         - Of which 2 did not match due to:
         Overload in function 'np_asarray': File: numba/np/arraymath.py: Line 4196.
           With argument(s): '(nestedarray(float64, (10,)))':
          Rejected as the implementation raised a specific error:
            LoweringError: Failed in nopython mode pipeline (step: native lowering)
          Can only insert double at [0] in [10 x double]: got i8*
          
          File "../../../miniconda3/envs/aesara-numba/lib/python3.9/site-packages/numba/np/arraymath.py", line 4208:
                      def impl(a, dtype=None):
                          return a
                          ^
          
          During: lowering "return $4return_value.1" at /home/anna/miniconda3/envs/aesara-numba/lib/python3.9/site-packages/numba/np/arraymath.py (4208)
     raised from /home/anna/miniconda3/envs/aesara-numba/lib/python3.9/site-packages/numba/core/errors.py:786
   
   During: resolving callee type: Function(<function asarray at 0x7fb0b026a040>)
   During: typing of call at /home/anna/miniconda3/envs/aesara-numba/lib/python3.9/site-packages/numba/np/arrayobj.py (1987)
   
   
   File "../../../miniconda3/envs/aesara-numba/lib/python3.9/site-packages/numba/np/arrayobj.py", line 1987:
       def impl(a):
           return np.asarray(a).shape
           ^

  raised from /home/anna/miniconda3/envs/aesara-numba/lib/python3.9/site-packages/numba/core/typeinfer.py:1074

During: resolving callee type: Function(<function shape at 0x7fb0b028e430>)
During: typing of call at /tmp/ipykernel_17638/1962980170.py (14)


File "../../../../../tmp/ipykernel_17638/1962980170.py", line 14:
<source missing, REPL/exec in use?>
```

</details>

**Comment by sklam** [comment]
Alsp, Numba should also not have triggered the error from:

```pytb
  File "/root/miniconda3/envs/numba_dev/lib/python3.9/site-packages/llvmlite/ir/builder.py", line 980, in insert_value
    instr = instructions.InsertValue(self.block, agg, value, idx, name=name)
  File "/root/miniconda3/envs/numba_dev/lib/python3.9/site-packages/llvmlite/ir/instructions.py", line 660, in __init__
    raise TypeError("Can only insert %s at %r in %s: got %s"
TypeError: Can only insert double at [0] in [10 x double]: got i8*
```

**Comment by gmarkall** [comment]
I wonder if https://github.com/numba/numba/pull/7359 fixes this

**Comment by gmarkall** [comment]
With #7359, the following:

```python
import numba
import numpy as np

dtype = np.dtype([
    ("a", np.float64, (10,))
])

data = np.zeros((), dtype=dtype)[()]

@numba.njit
def foo(data):
    return np.shape(data.a)

print(foo(data))
```

prints

```
(10,)
```

So I think this will be resolved by #7359.

**Comment by gmarkall** [comment]
@astoeriko Are there any similar use cases that ought to be checked? Can you try with #7359 and check if it works for all your intended use cases?

**Comment by astoeriko** [comment]
> So I think this will be resolved by #7359.

That sounds great, thanks for testing! I wanted to test it for a while but did not actually get around to do so.
> Are there any similar use cases that ought to be checked? Can you try with #7359 and check if it works for all your intended use cases?

I will try with my use cases and get back to you.
